crates/sui-core/src/checkpoints/checkpoint_executor/utils.rs

```diff
@@ -174,6 +175,75 @@ pub(super) fn assert_not_forked(
     }
 }
 
+pub(super) fn assert_checkpoint_not_forked(
+    locally_built_checkpoint: &CheckpointSummary,
+    verified_checkpoint: &VerifiedCheckpoint,
+    checkpoint_store: &CheckpointStore,
+) {
+    assert_eq!(
+        locally_built_checkpoint.sequence_number(),
+        verified_checkpoint.sequence_number(),
+        "Checkpoint sequence numbers must match"
+    );
+
+    if locally_built_checkpoint.digest() == *verified_checkpoint.digest() {
+        return;
+    }
+
+    let verified_checkpoint_summary = verified_checkpoint.data();
+
+    if locally_built_checkpoint.content_digest == verified_checkpoint_summary.content_digest {
+        // fork is in the checkpoint header
+        fatal!("Checkpoint fork detected in header! Locally built checkpoint: {:?}, verified checkpoint: {:?}",
+            locally_built_checkpoint,
+            verified_checkpoint
+        );
+    } else {
+        let local_contents = checkpoint_store
+            .get_checkpoint_contents(&locally_built_checkpoint.content_digest)
+            .expect("db error")
+            .expect("contents must exist if checkpoint was built locally!");
+
+        let verified_contents = checkpoint_store
+            .get_checkpoint_contents(&verified_checkpoint_summary.content_digest)
+            .expect("db error")
+            .expect("contents must exist if checkpoint has been synced!");
+
+        // fork is in the checkpoint contents
+        let mut local_contents_iter = local_contents.iter();
```

Comment:
can this be simplified with itertools::zip_eq

Reply:
zip_eq would just immediately panic if they are different lengths so i wouldn't have a chance to print out a useful message